src/types.jl

```diff
@@ -0,0 +1,156 @@
+
+"""SuperClass of all Prometheus Metrics."""
+abstract type PromMetric end
+
+"""To be implemented."""
+abstract type SummaryMetric <: PromMetric end
+
+"""
+    CounterMetric
+
+Struct to hold a prometheus counter metric.
+
+## Usage:
+- Values must all be the same numeric type, defined by vtype upon struct creation.
+- Label Keys should be pre-defined in label_keys before metrics are added. Modifying labels during runtime is NOT encouraged.
+- Labels are order Sensitive, when adding metrics.
+- Outputting timestamps are optional, therefore in order to retrieve timestamp log_timestamp kwarg must be set to True.
+
+## Notes:
+- Dict keys in label_data are implemented as sets of paired strings, look like this:
+```
+julia> a = Set(["a"=>"b", "c"=>"d"])
+julia> b[a] = 10
+julia> b[Set(["c"=>"d","a"=>"b"])] == b[Set(["a"=>"b","c"=>"d"])]  # true
+````
+"""
+struct CounterMetric{VT <: Number, LT<: Base.AbstractLock} <: PromMetric
+    name::String                                            # Required for TYPE string
+    desc::String                                            # Description for HELP string, can be blank str
+    label_keys::Vector{String}                              # Ordered List of Possible Keys for use in labels
+    label_data::Dict{Set{Pair{String, String}},VT}          # Dict that stores metric data per label
+    label_timestamp::Dict{Set{Pair{String, String}},Int}    # Dict that stores timestamp data
+    log_timestamp::Bool                                     # Field to check if the label_timestamp field needs to be updated
+    _vtype::Type                                            # for storing the value type, useful for checking
+    _lk::LT                                                 # lock
+end
+CounterMetric(n; t::Bool=false, vtype::Type=Int) = CounterMetric(n, "", Vector{String}(); t=t, vtype=vtype)
+CounterMetric(n, d; t::Bool=false, vtype::Type=Int) = CounterMetric(n, d, Vector{String}(); t=t, vtype=vtype)
+function CounterMetric(n, d, labels::Vector{String}; t::Bool=false, vtype::Type=Int) 
+    cm = CounterMetric(n, d, labels, Dict{Set{Pair{String, String}}, vtype}(), Dict{Set{Pair{String, String}}, Int}(), t, vtype, Threads.ReentrantLock())
+    add_metric_to_collector!(cm)  # always add to default collector
```

Comment:
Yes - I did remember right, we are already adding the metric to the default collector by default, so no need to have that step in the readme. Unless you want to change the example in the readme to adding to a custom collector.